homeassistant/components/sensor/yr.py

```diff
@@ -63,32 +69,25 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
 
     coordinates = dict(lat=latitude, lon=longitude, msl=elevation)
 
-    weather = YrData(coordinates)
-
     dev = []
     for sensor_type in config[CONF_MONITORED_CONDITIONS]:
-        dev.append(YrSensor(sensor_type, weather))
+        dev.append(YrSensor(sensor_type))
 
-    # add symbol as default sensor
-    if len(dev) == 0:
-        dev.append(YrSensor("symbol", weather))
-    add_devices(dev)
+    weather = YrData(hass, coordinates, dev)
+    tasks = [async_add_devices(dev), weather.async_update()]
+    yield from asyncio.gather(*tasks, loop=hass.loop)
```

Comment:
Instead of this work around you can just write

```python
yield from asyncio.gather(async_add_devices(dev), weather.async_update(), loop=hass.loop)
```

However, in this case it is actually important that `async_add_devices` runs first, because it sets the `hass` attribute and entity_id which is needed to call `async_update_ha_state`. However, without `weather.async_update`, you will write `unknown` to the state machine.

Do note that this way you will first get a bunch of unknown states that will resolve once `weather.async_update` comes through.

I guess the ultimate solution would be to make an async friendly throttle.